File: javascript/Data-Structures/Tree/SegmentTree.js
class SegmentTree {
  size
  tree
  constructor(arr) {
    
    this.size = arr.length
    this.tree = new Array(2 * arr.length)
    this.tree.fill(0)

    this.build(arr)
  }

  build(arr) {
    const { size, tree } = this
    
    for (let i = 0; i < size; i++) {
      tree[size + i] = arr[i]
    }

    for (let i = size - 1; i > 0; --i) {
      
      tree[i] = tree[i * 2] + tree[i * 2 + 1]
    }
  }

  update(index, value) {
    const { size, tree } = this

    index += size
    
    tree[index] = value

    for (let i = index; i > 1; i >>= 1) {
      
      tree[i >> 1] = tree[i] + tree[i ^ 1]
    }
  }

  query(left, right) {
    const { size, tree } = this
    
    right++
    let res = 0

    for (left += size, right += size; left < right; left >>= 1, right >>= 1) {
      
      if ((left & 1) > 0) {
        res += tree[left++]
      }

      if ((right & 1) > 0) {
        res += tree[--right]
      }
    }

    return res
  }
}

export { SegmentTree }

File: csharp/DataStructures/SegmentTrees/SegmentTree.cs
using System;

namespace DataStructures.SegmentTrees;

public class SegmentTree
{
    
    public SegmentTree(int[] arr)
    {
        
        var pow = (int)Math.Pow(2, Math.Ceiling(Math.Log(arr.Length, 2)));
        Tree = new int[2 * pow];

        Array.Copy(arr, 0, Tree, pow, arr.Length);

        for (var i = pow - 1; i > 0; --i)
        {
            Tree[i] = Tree[Left(i)] + Tree[Right(i)];
        }
    }

    public int[] Tree { get; }

    public int Query(int l, int r) =>
        Query(++l, ++r, 1, Tree.Length / 2, 1);

    protected int Right(int node) => 2 * node + 1;

    protected int Left(int node) => 2 * node;

    protected int Parent(int node) => node / 2;

    protected virtual int Query(int l, int r, int a, int b, int i)
    {
        
        if (l <= a && b <= r)
        {
            return Tree[i];
        }

        if (r < a || b < l)
        {
            
            return 0;
        }

        var m = (a + b) / 2;

        return Query(l, r, a, m, Left(i)) + Query(l, r, m + 1, b, Right(i));
    }
}
